python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/asset_daemon_scenario.py

```diff
@@ -0,0 +1,246 @@
+import datetime
+import logging
+from typing import (
+    Any,
+    Callable,
+    Iterable,
+    NamedTuple,
+    Optional,
+    Sequence,
+    Tuple,
+    Type,
+)
+
+import dagster._check as check
+import pendulum
+from dagster import (
+    AssetKey,
+    AssetsDefinition,
+    AssetSpec,
+    AutoMaterializePolicy,
+    DagsterInstance,
+    RunRequest,
+    asset,
+)
+from dagster._core.definitions import materialize
+from dagster._core.definitions.asset_daemon_context import AssetDaemonContext
+from dagster._core.definitions.asset_daemon_cursor import AssetDaemonCursor
+from dagster._core.definitions.asset_graph import AssetGraph
+from dagster._core.definitions.auto_materialize_rule import (
+    AutoMaterializeAssetEvaluation,
+    AutoMaterializeRule,
+    AutoMaterializeRuleEvaluation,
+    AutoMaterializeRuleEvaluationData,
+)
+from dagster._core.definitions.events import CoercibleToAssetKey
+
+
+class AssetRuleEvaluationSpec(NamedTuple):
+    rule: AutoMaterializeRule
+    partitions: Optional[Sequence[str]] = None
+    rule_evaluation_data: Optional[AutoMaterializeRuleEvaluationData] = None
+
+    def with_rule_evaluation_data(
+        self, data_type: Type[AutoMaterializeRuleEvaluationData], **kwargs
+    ) -> "AssetRuleEvaluationSpec":
+        # convert sets of CoercibleToAssetKey to frozensets of AssetKey (for convenience)
+        transformed_kwargs = {
+            key: frozenset(AssetKey.from_coercible(v) for v in value)
+            if isinstance(value, set)
+            else value
+            for key, value in kwargs.items()
+        }
+        return self._replace(
+            rule_evaluation_data=data_type(**transformed_kwargs),
+        )
+
+    def resolve(self) -> Tuple[AutoMaterializeRuleEvaluation, Optional[Sequence[str]]]:
+        return (
+            AutoMaterializeRuleEvaluation(
+                rule_snapshot=self.rule.to_snapshot(),
+                evaluation_data=self.rule_evaluation_data,
+            ),
+            self.partitions,
+        )
+
+
+class AssetDaemonScenarioState(NamedTuple):
+    asset_specs: Sequence[AssetSpec]
+    current_time: datetime.datetime = pendulum.now()
+    run_requests: Sequence[RunRequest] = []
+    cursor: AssetDaemonCursor = AssetDaemonCursor.empty()
+    evaluations: Sequence[AutoMaterializeAssetEvaluation] = []
+    logger: logging.Logger = logging.getLogger("dagster.amp")
+    internal_instance: Optional[DagsterInstance] = None
```

Comment:
What does "internal" mean in this context?

Reply:
If NamedTuples were allowed to have properties that started with an underscore, I'd just name this "_instance". Basically, I didn't want to have to pass in a dummy instance every time we created one of these (so ended up with type Optional[DagsterInstance] and defaulting to None), but at the same time, wanted to have an `instance` property that was of type DagsterInstance (as I know this will always get set by the time it comes to access it). So ended up with this slightly awkward situation.

Renamed to "scenario_instance"

Reply:
If we went the non-lambda approach, would this be necessary? The instance could just be provided by the test framework at the time that all the actions are evaluated.

Reply:
I guess more specifically, we'd have something like:

```python
class AssetDaemonScenario(NamedTuple):
    ...
    pending_actions: Sequence[Callable[[AssetDaemonScenario, DagsterInstance], AssetDaemonScenario]

    def with_runs(self, *run_requests: RunRequest) -> AssetDaemonScenario:
        def fn(scenario, instance):
            ... # do stuff to add runs
        return self._replace(pending_actions=self.pending_actions + [fn])

```

So definitely does clean up things on the caller's side of the street, but makes things a bit messy on the framework side.